### PR TITLE
perf: apply left side predicate pushdown also to right side on semi join

### PR DESCRIPTION
**Pull request by gab23r**
linked to #10841

Added the semi join into the list of joins that can accept this predicate pushdown optimization.
Also remove the TODO because no other join can benefit from it.

### CODE REVIEW

File: crates/polars-plan/src/logical_plan/optimizer/predicate_pushdown/join.rs [hunk 1]
```diff
@@ -133,12 +133,19 @@ pub(super) fn process_join(
...
         {
             insert_and_combine_predicate(&mut pushdown_right, predicate, expr_arena);
             filter_right = true;
         }
+        #[cfg(feature = "semi_anti_join")]
+        if filter_left
+            && all_pred_cols_in_left_on(predicate, expr_arena, &left_on)
+            && matches!(&options.args.how, JoinType::Semi)
+        {
+            insert_and_combine_predicate(&mut pushdown_right, predicate, expr_arena);
```
Comment (ritchie46) [inline crates/polars-plan/src/logical_plan/optimizer/predicate_pushdown/join.rs:146]: Now we do an extra traversal. Can you rewrite the `matches!` on line 136 to a proper `match` statement where we put the Semi as well?

Comment (gab23r) [reply crates/polars-plan/src/logical_plan/optimizer/predicate_pushdown/join.rs:146]: That's what I did in my initial commit but I didn't get where to put the feature gate. Can I put it inside the if statement ?

Comment (ritchie46) [reply crates/polars-plan/src/logical_plan/optimizer/predicate_pushdown/join.rs:146]: Something like this:

```rust
match how {
  #[cfg(feature = ..)]
  JoinType::Semi,
  JoinType::Left
  ...
}
```

**Comment by gab23r** [comment]
In the previous code, it was possible to have filter_right = true and block_pushdown_right = true. This seems not correct. I should fix this as well, or I miss something ?

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**